core/src/test/java/greencity/controller/HabitControllerTest.java

```diff
@@ -0,0 +1,290 @@
+package greencity.controller;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+import static org.springframework.http.MediaType.APPLICATION_JSON;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+import greencity.GreenCityApplication;
+import greencity.dto.PageableDto;
+import greencity.dto.habit.AddCustomHabitDtoRequest;
+import greencity.dto.habit.AddCustomHabitDtoResponse;
+import greencity.dto.habit.HabitDto;
+import greencity.dto.shoppinglistitem.ShoppingListItemDto;
+import greencity.dto.user.UserProfilePictureDto;
+import greencity.dto.user.UserVO;
+import greencity.exception.exceptions.BadRequestException;
+import greencity.service.HabitService;
+import greencity.service.LanguageService;
+import greencity.service.TagsService;
+import greencity.service.UserService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.modelmapper.ModelMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.data.domain.Pageable;
+import org.springframework.http.MediaType;
+import org.springframework.mock.web.MockMultipartFile;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.web.multipart.MultipartFile;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.*;
+
+@WithMockUser(username = "user")
+@WebMvcTest(HabitController.class)
+@ContextConfiguration(classes = {GreenCityApplication.class})
+class HabitControllerTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private ModelMapper modelMapper;
+
+    @MockBean
+    private LanguageService languageService;
+
+    @MockBean
+    private UserService userService;
+
+    @MockBean
+    private TagsService tagsService;
+
+    @MockBean
+    private HabitService habitService;
+
+    private UserVO userVO;
+
+    @BeforeEach
+    void setUp() {
+        userVO = UserVO.builder()
```

Comment:
there is a test user in ModelUtils.getUser()

Reply:
I think the idea is not bad since a test class with a similar model already exists. However, in this case, I think it would be appropriate to move all models to static methods in the ModelUtils class for project consistency. Yet, I also do not believe that moving model creation into a static method in another class is a good idea because it increases code readability time, especially during reviews. Also, many tests might require special data in the fields. If we consider this class as a way to simplify the work of tests for mappers, then the reliability of PR checks becomes extremely low.

However, speaking about the context of this project, it might be permissible, but still, wherever possible, I would advise avoiding it.

The tests that we wrote in the mappers using this class are difficult to read and unpredictable for the reader. This is only if they go to the branch itself and check it, which they simply may not have time for.